Flatpak puller

nintyfan opened the pull request:
Some changes need to be done, like:
1. Using calamares utilities to deal with creating processes and handling output data from it
2. Eliminating duplicates

Currently, it is speedy, have small amount of code and works. I must told, it is hacky. I check if flatpak write to output line contains dot to filter-out broken data (empty lines?): I did not check, what these data was yet, so this is third point to correct. Without that check, netinstall displays virtual packages (with empty names).

Commit:
Made possible to select flatpaks to install (fill list of netinstall)

Commit:
Made possible to select flatpaks to install (fill list of netinstall)

Commit:
- User must select flatpak to install it

Commit:
Speedup operations

Commit:
Skipping wrong package names (blank lines)

nintyfan commented:
https://github.com/calamares/calamares-extensions/issues/27

Commit:
Flatpaks will be correctly marked as installed or not

Commit:
Correctly set items as markable in netinstall

Commit:
Disallow to uncheck installed flatpaks

Commit:
Use CalamaresUtils to create process

Commit:
User see if flatpak was installed (and cannot be removed)

nintyfan commented:
First point was done.
Second was untouched.
Third is also untouched.

Many work was done, but from your point of view it could not, because of my description.

Commit:
User do not see duplicate names of packages

Commit:
Remove unnecessary check

Commit:
Remove unused variables

Commit:
Repair code style

Commit:
Repair code style

nintyfan commented:
Yes. I will do this tomorrow.

Commit:
Repair coding-style (apply clang-format)

Commit:
Repair coding-style (apply clang-format)

nintyfan commented:
> Can you apply the Calamares code format/style using `clang-format`

Done

dalto8 commented:
The commit where `clang-format` was applied appears to have removed all the code.

nintyfan commented:
> Hi @nintyfan,
> 
> I am currently working on installing a module which supports both Flatpak and apt packages. I have managed to download the necessary files and successfully compiled the C++ source code into a .so (shared object) file. However, I am unsure about how to properly utilize this module to achieve my goal.
> 
> Could you please guide me on how to use this file or integrate it within my existing setup? Any help or guidance would be greatly appreciated.
> 
> Thank you! This is my yaml:
> 
> ```
> - name: "I am a Light User"
>   description: "Installs the basics. Browser, Multimedia Player, FreeTube etc"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - libreoffice
> 
> - name: "I am a GAMER!"
>   description: "Installs everything needed for a Gaming Ready System"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - steam-installer
>     - gamemode
>     - goverlay
>     - gamescope
>     - wmctrl
>     - libreoffice
> and my idea put
>   flatpaks:
>     - com.expidusos.calculator 
> like packages. Is this possible? 
> ```

I have send PR to main calamares bugtracker. It contains support for installing flatpak. This code is designed to achieve different goal: populate list of packages. For more info, see related issue.

nintyfan commented:
> Hi @nintyfan,
> 
> I am currently working on installing a module which supports both Flatpak and apt packages. I have managed to download the necessary files and successfully compiled the C++ source code into a .so (shared object) file. However, I am unsure about how to properly utilize this module to achieve my goal.
> 
> Could you please guide me on how to use this file or integrate it within my existing setup? Any help or guidance would be greatly appreciated.
> 
> Thank you! This is my yaml:
> 
> ```
> - name: "I am a Light User"
>   description: "Installs the basics. Browser, Multimedia Player, FreeTube etc"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - libreoffice
> 
> - name: "I am a GAMER!"
>   description: "Installs everything needed for a Gaming Ready System"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - steam-installer
>     - gamemode
>     - goverlay
>     - gamescope
>     - wmctrl
>     - libreoffice
> and my idea put
>   flatpaks:
>     - com.expidusos.calculator 
> like packages. Is this possible? 
> ```

I have send PR to main calamares bugtracker. It contains support for installing flatpak. This code is designed to achieve different goal: populate list of packages. For more info, see related issue.

> Hi @nintyfan,
> 
> I am currently working on installing a module which supports both Flatpak and apt packages. I have managed to download the necessary files and successfully compiled the C++ source code into a .so (shared object) file. However, I am unsure about how to properly utilize this module to achieve my goal.
> 
> Could you please guide me on how to use this file or integrate it within my existing setup? Any help or guidance would be greatly appreciated.
> 
> Thank you! This is my yaml:
> 
> ```
> - name: "I am a Light User"
>   description: "Installs the basics. Browser, Multimedia Player, FreeTube etc"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - libreoffice
> 
> - name: "I am a GAMER!"
>   description: "Installs everything needed for a Gaming Ready System"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - steam-installer
>     - gamemode
>     - goverlay
>     - gamescope
>     - wmctrl
>     - libreoffice
> and my idea put
>   flatpaks:
>     - com.expidusos.calculator 
> like packages. Is this possible? 
> ```

I think you should place source field to package group. Next - packaging module should split packages based on this info and install it separately. To achieve this, netinstall or other should copy source info from package group into related packages.

adriaandegroot commented on modules/flatpakinfo/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+# === This file is part of Calamares - <https://calamares.io> ===
+#
+#   SPDX-FileCopyrightText: 2020 Adriaan de Groot <groot@kde.org>
```
No it's not.

adriaandegroot commented on modules/flatpakinfo/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+        flatpakInfoJob.h
+        ItemFlatpak.h
+        PackagePool.h
+        flatpakInfoJob.cpp
```
Weirdly inconsistent file-names

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <QVariantMap>
```
Missing copyright header. Includes all mixed together (Qt, STL, C, Calamares).

nintyfan replied:
I check in [nintyfan:FlatpakPuller](https://github.com/nintyfan/calamares-extensions/tree/FlatpakPuller) and changes are there, but cannot found it here. Could someone's else check if changes are ready to apply. I talk about licensing information.

nintyfan replied:
Sorry for mess up. When seen last commit, changes you proposed are applied. But when clicking on title header, I got message that this commit do not belong to any branch. Could you see last commit in current PR?

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.cpp:
```diff
@@ -0,0 +1,64 @@
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include <ext/stdio_filebuf.h>
```
Why? Actually that goes for all the STL / C headers -- what do ytou need them for?

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.cpp:
```diff
@@ -0,0 +1,64 @@
+
+#include "utils/CalamaresUtilsSystem.h"
+
+QStringList installed;
+bool installedFilled;
```
Global variables? Also, UB if `installedFilled` is touched before something writes to it -- but does it mean anything other than `!installed.isEmpty()`?

nintyfan replied:
> Global variables? Also, UB if `installedFilled` is touched before something writes to it -- but does it mean anything other than `!installed.isEmpty()`?

This not the same. I prefer to avoid operation with long duration. This is only 1/8-4 byte of memory and we avoid calling flatpak often. Imagine there is not flatpaks installed at all and integrator adds only way to install flatpaks (without adding any to distribution by default).

nintyfan replied:
Done.

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+void
+ItemFlatpak_freeMem( void )
```
CamelCase or snake_case, pick one, not both.

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.cpp:
```diff
@@ -0,0 +1,64 @@
+{
+    QString line;
+    auto process = CalamaresUtils::System::instance()->targetEnvCommand(
+        QStringList { QString::fromStdString( "flatpak" ),
```
Unnecessary extra conversion to std::string

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.cpp:
```diff
@@ -0,0 +1,64 @@
+                      QString::fromStdString( "--app" ),
+                      QString::fromStdString( "--columns=application" ) } );
+    auto output_str = process.second;
+    QTextStream output( &output_str );
```
There is surely a better way to get output line-by-line from a command.

nintyfan replied:
I do implementation thanks to QString::indexOf and QString::mid . There  could be nicer, but slower, thanks to QString::section .

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.cpp:
```diff
@@ -0,0 +1,64 @@
+    }
+    // check if it is installed
+    PackageItem item( CalamaresUtils::getString( item_map, "appstream" ) );
+    item.setInstalled( false );
+
+    for ( auto names : installed )
+    {
+        if ( names == CalamaresUtils::getString( item_map, "appstream" ) )
+        {
+            item.setInstalled( true );
+        }
+    }
```
`item.setInstalled(installed.contains(...)` where you probably want to get the name just once (the *appstream* key from the map)

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.h:
```diff
@@ -0,0 +1,38 @@
+/* === This file is part of Calamares - <https://calamares.io> ===
+ *
+ *   SPDX-FileCopyrightText: 2019 Adriaan de Groot <groot@kde.org>
```
No it's not.

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.h:
```diff
@@ -0,0 +1,38 @@
+    bool getInstalled( void ) { return this->installed; }
+
+private:
+    QString appstreamid;
```
Most data members in Calamares code have a `m_` prefix.

nintyfan replied:
What do you mean?

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.h:
```diff
@@ -0,0 +1,38 @@
+        , installed( false )
+    {
+    }
+    QString& getAppStreamId( void ) { return this->appstreamid; }
```
`this->` is almost never necessary (and here it definitely is not needed)

adriaandegroot commented on modules/flatpakinfo/ItemFlatpak.h:
```diff
@@ -0,0 +1,38 @@
+    }
+    QString& getAppStreamId( void ) { return this->appstreamid; }
+    void setInstalled( bool installed ) { this->installed = installed; }
+    bool getInstalled( void ) { return this->installed; }
```
`(void)` should be `()`, this is not C

adriaandegroot replied:
.. and getters should be `const`, too.

adriaandegroot commented on modules/flatpakinfo/PackagePool.h:
```diff
@@ -0,0 +1,15 @@
+/* === This file is part of Calamares - <https://calamares.io> ===
+ *
+ *   SPDX-FileCopyrightText: 2021 Evan James <dalto@fastmail.com>
```
No it's not.

adriaandegroot commented:
So, overall, here is the thing you are trying to do:
- get a list of flatpaks (that are available for installation) by running flatpak tools inside the target system
- adding those flatpaks to the data for netinstall, so that a following netinstall step can display those packages
- (and then existing mechanisms will somehow install those flatpaks that are selected in netinstall)

dalto8 commented:
> I do not known if mechanism inside packages is merged.

That was merged on August 12th.

Commit:
Add properl licensing info

Commit:
Make filenames more consistend

Commit:
Make filenames more consistend

Commit:
Cleanup includes in ItemFlatpak

Commit:
Remove redundant this

Commit:
Remove unnecessary void

Commit:
Allow to better optimise getters

Commit:
Use Camel-Case

Commit:
Avoid conversion to std::string

Commit:
Change implementation of preparing installed package list routine

Commit:
Make code more functional by eliminating loop

Commit:
Repair data members' names

Commit:
Initialize installed list on startup of module

dalto8 commented:
For me, personally, I would not be comfortable merging this with that global variable and all those free functions that reference it.  I really think it should have an object model more similar to the rest of the C++ Calamares code.

However, this is extensions so someone else may feel differently.

nintyfan commented:
Which global variable? I remove boolean variable. Did you write about
installed package list? In this case, I will move it to module class, or at
least - try.

niedz., 22 paź 2023, 17:08 użytkownik dalto8 ***@***.***>
napisał:

> For me, personally, I would not be comfortable merging this with that
> global variable and all those free functions that reference it. I really
> think it should have an object model more similar to the rest of the C++
> Calamares code.
>
> However, this is extensions so someone else may feel differently.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/calamares/calamares-extensions/pull/28#issuecomment-1774119126>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGXA6Q353WJR4KTV45CKLBTYAUZGRAVCNFSM6AAAAAA4L43CVGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONZUGEYTSMJSGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

dalto8 commented:
> Which global variable?

Any global variable.



> Did you write about installed package list?

Yes, that was the one I noticed.

Commit:
Remove global variable installed list

nintyfan commented:
Variable removed.
I am not object-oriented programming enthusiast, so maybe I do something wrong, especially look at constructor I create in latest patch: is it valid to prepare installed list inside constructor? Maybe better to not establish connection and allocate memory inside and move this to separate method?

Commit:
Remove global variable packages

nintyfan commented:
I think, there is no more global-variables in flatpakinfo module.

nintyfan committed:
Merge branch 'calamares:calamares' into FlatpakPuller

nintyfan commented:
> Hi @nintyfan,
> 
> I am currently working on installing a module which supports both Flatpak and apt packages. I have managed to download the necessary files and successfully compiled the C++ source code into a .so (shared object) file. However, I am unsure about how to properly utilize this module to achieve my goal.
> 
> Could you please guide me on how to use this file or integrate it within my existing setup? Any help or guidance would be greatly appreciated.
> 
> Thank you! This is my yaml:
> 
> ```
> - name: "I am a Light User"
>   description: "Installs the basics. Browser, Multimedia Player, FreeTube etc"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - libreoffice
> 
> - name: "I am a GAMER!"
>   description: "Installs everything needed for a Gaming Ready System"
>   critical: false
>   hidden: false
>   selected: false
>   packages:
>     - steam-installer
>     - gamemode
>     - goverlay
>     - gamescope
>     - wmctrl
>     - libreoffice
> and my idea put
>   flatpaks:
>     - com.expidusos.calculator 
> like packages. Is this possible? 
> ```

Hi. I tries to send private message, but I cannot. You can use instance of Calamares module and joins PACKAGES string with instance name - both in packages.py and in some file of calamares library (search for add packages to install), where packages list to install was built.

To get instance:
netinstall@normal_packages
packages@normal_packages
netinstall@flatpak
packages@flatpak

You must also wrote special configuration for each instance of modules (netinstall@normal_packages, packages@normal_packages, ...) . I think, that it should work.

Airmancooma commented:
> > Hi @nintyfan,
> > I am currently working on installing a module which supports both Flatpak and apt packages. I have managed to download the necessary files and successfully compiled the C++ source code into a .so (shared object) file. However, I am unsure about how to properly utilize this module to achieve my goal.
> > Could you please guide me on how to use this file or integrate it within my existing setup? Any help or guidance would be greatly appreciated.
> > Thank you! This is my yaml:
> > ```
> > - name: "I am a Light User"
> >   description: "Installs the basics. Browser, Multimedia Player, FreeTube etc"
> >   critical: false
> >   hidden: false
> >   selected: false
> >   packages:
> >     - libreoffice
> > 
> > - name: "I am a GAMER!"
> >   description: "Installs everything needed for a Gaming Ready System"
> >   critical: false
> >   hidden: false
> >   selected: false
> >   packages:
> >     - steam-installer
> >     - gamemode
> >     - goverlay
> >     - gamescope
> >     - wmctrl
> >     - libreoffice
> > and my idea put
> >   flatpaks:
> >     - com.expidusos.calculator 
> > like packages. Is this possible? 
> > ```
> 
> Hi. I tries to send private message, but I cannot. You can use instance of Calamares module and joins PACKAGES string with instance name - both in packages.py and in some file of calamares library (search for add packages to install), where packages list to install was built.
> 
> To get instance: netinstall@normal_packages packages@normal_packages netinstall@flatpak packages@flatpak
> 
> You must also wrote special configuration for each instance of modules (netinstall@normal_packages, packages@normal_packages, ...) . I think, that it should work.

Oh sorry I completely forgot about that but thanks for replying! I was so annoyed with calamares that I made my own in python for my debian linux. It still needs some work because systemd-boot and grub need to be separate etc etc but since I do it I know what I need and I can configure what I want a thousand times better than in calamares where there is no proper description...

nintyfan committed:
Update ItemFlatpak.cpp

Commit:
CMakeLists: Copy KF CoreAddons and FeatureSummary imports from Calamares

Without these the package no-longer compiles with recent versions of
cmake.

Commit:
Port modules to Calamares 3.3.0

Calamares 3.3.0 changed the import name and namespace of
CalamaresUtilsSystem and CalamaresUtilsGui.

adriaandegroot committed:
CMake: find Calamares first, then other bits

adriaandegroot committed:
CMake: find Calamares first, then other bits

adriaandegroot committed:
CMake: bump version and requirements

Builds cleanly with Calamares 3.3.1

adriaandegroot committed:
CMake: bump version and requirements

Builds cleanly with Calamares 3.3.1

adriaandegroot committed:
CMake: be more verbose when looking for Calamares

adriaandegroot committed:
CHANGES: post-release housekeeping

nintyfan committed:
(Allow to work with newer version)

nintyfan committed:
Merge branch 'calamares:calamares' into FlatpakPuller

nintyfan reviewed (commented):
Probably done all of requests.
